### ISSUE TITLE
Release 4.3.0 - Release Candidate 4 - AWS Module testing

### ISSUE THREAD

**Issue by CarlosRS9** [wazuh/wazuh#12183]
# AWS module testing

## Modules tests information
|                                      |                                                |
|--------------------------------------|------------------------------------------------|
| **Main module tests issue**          | #12181                                         |
| **Version**                          | 4.3.0                                          |
| **Release candidate #**              | RC4                                            |
| **Tag**                              | https://github.com/wazuh/wazuh/tree/v4.3.0-rc4 |
| **Previous modules tests issue**     | #11857                                         |

## Test report procedure

All test results must have one the following statuses: 
|                 |                                                                         |
|-----------------|-------------------------------------------------------------------------|
| :green_circle: | The test passed successfully. 										    |
| :yellow_circle: | The test passed but some enhancements need to be applied to the module. |
| :red_circle:    | The test failed.                                                        |
| :white_circle:  | The test does not apply.                                                |

Any failing test must be properly addressed with a new issue, detailing the error and the possible cause. It must be included in the [Release 4.3.0 project](https://github.com/orgs/wazuh/projects/4/) for triage.

Any expected fail or skipped test must have an issue justifying the reason. All auditors must validate the justification for an expected fail or skipped test.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

Every test will be performed on a Manager unless otherwise specified.

## Test results

The `only_logs_after` parameter was the only functionality tested for each bucket/service type on this RC because it is the only parameter that has seen changes with respect to the previous RC and at the same time it is the best way to test the basic functionality of the module.
Check RC2 and RC3 for more details about the previously performed tests whose results are still relevant.

| Bucket/Service type | Test result    | Issue ref. |
|---------------------|----------------|------------|
| Umbrella            | :green_circle: | -          |
| Inspector           | :green_circle: | -          |
| CloudTrail          | :green_circle: | -          |
| VPC                 | :green_circle: | -          |
| Config              | :red_circle:   | #12245     |
| LB                  | :green_circle: | -          |
| Macie               | :green_circle: | -          |
| GuardDuty           | :green_circle: | -          |
| WAF                 | :green_circle: | -          |
| Server Access       | :green_circle: | -          |

The following test was performed to ensure the module worked as expected in a Wazuh Agent installation:

|   | Test result | Issue ref. |
|---|-------------| - |
| Test the module in an agent | :green_circle: | - |

### Conclusions
We opened #12245 because we detected that the `Config` module could generate a wrong marker under some circumstances. This won't necessarily generate duplicate alerts, but it might happen so the error must be fixed soon.

## Auditors validation
The definition of done for this one is the validation of the conclusions and the test results from all auditors.

All checks from below must be accepted in order to close this issue.

- [ ] @davidjiglesias
- [x] @CarlosRS9
- [x] @noise-kngdm

**Comment by CarlosRS9** [comment]
# Umbrella

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Test only_logs_after |  |  |
| Test iplogs   |   | - |
| Test dnslogs   |   | - |
| Test proxylogs   |   | - |

### Conclusions
The module is not working adequately when executing it for the first time without spceifying a value for the `only_logs_after` parameter. #11899 was opened to fix that.

### Test details

#### Test Umbrella
Since the three types of Cisco Umbrella buckets use the same underlying logic, we will only need to test that the `only_logs_after` parameter for one of them.

Remove the `/var/ossec/wodles/aws/s3_cloudtrail.db` file and run the module using the following command to obtain every log from the Umbrella bucket for the given date.

##### First execution without specifying an only_logs_after value
Upload one log file to the bucket with the date of execution. Then execute the command. It should use as marker an string with the date of execution and collect the uploaded log.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

###### Conclusions


##### First execution specifying an only_logs_after value set in the past
Remove the database file and run the command the using `2021-Dec-20` date as `only_logs_after`. 3 logs should be collected and sent to analysisd.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -s 2021-Dec-20 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs have been processed.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -s 2021-Dec-20 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

###### Conclusions


##### Execute the command again with an only_logs_after value set earlier in the past

Run the module again, keeping the database file, using the `2021-Dec-01` date as `only_logs_after`. It should use as marker an string with the date of execution and not consider the `only_logs_after` value.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -s 2021-Dec-01 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

###### Conclusions


##### Execute the module having a last key set and no only_logs_after value
Without removing the database, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from cisco_umbrella where created_date in (select max(created_date) from cisco_umbrella);'`. This command will remove the more recents logs. Then execute the module without specifying an `only_logs_after` value. It should only collect the previously uploaded log.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

###### Conclusions


##### Execute the module having an only_logs_after later than the last key
The module should only fetch the logs corresponding to the days set by the only_logs_after date onwards, and it should skip the logs in between. 1 log should be processed.
Keeping the database file, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from cisco_umbrella where created_date not in (select min(created_date) from cisco_umbrella);'`. 

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -p dev -d2 -s YYYY-MMM-DD
```
Being `YYYY-MMM-DD` the data corresponding to the date of execution.

<details>
    <summary>Command output</summary>


</details>

###### Conclusions


##### Execute the module having an only_logs_after earlier than the last key
Without removing the database, execute `sqlite3 /var/ossec/wodles/aws/s3_cloudtrail.db 'delete from cisco_umbrella where created_date not in (select min(created_date) from cisco_umbrella);'`. This command will remove all the logs but the older ones. The execution of the module must result in collecting the last 3 logs, using as marker the last key.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -p dev -d2 -s 2021-Dec-01 
```

<details>
    <summary>Command output</summary>


</details>

###### Conclusions


##### Conclusions of the only_logs_after tests
We found a bug when executing the module for the first time without specifying an `only_logs_after` value, and opened #11899 to fix that.

#### Test iplogs
This command should only fetch `iplogs`.
Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/iplogs -t cisco_umbrella -s 2021-Dec-20 -p dev -d2
```

<details>
    <summary>Command output</summary>


</details>

##### Conclusions


#### Test proxylogs
This command should only fetch `proxylogs`.
Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/proxylogs -t cisco_umbrella -s 2021-Dec-15 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


#### Test dnslogs
This command should only fetch `dnslogs`.
Use the command:
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-umbrella -l 166157441623/dnslogs -t cisco_umbrella -s 2021-Dec-15 -p dev -d2
```

<details>
    <summary>Command output</summary>



</details>

##### Conclusions

**Comment by CarlosRS9** [comment]
# Inspector

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events was sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Using only_logs_after |  |  |
| Using reparse |  | |


## Test details
Before starting these tests, at least three assessments runs must have been run in three different dates.

### Test only_logs_after

Remove the `/var/ossec/wodles/aws/aws_services.db` file and run the module using the following command to obtain every finding log group for the given date.

#### First execution without specifying an only_logs_after value

Use the command: 
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -p dev -r us-east-1
```

Unless a new assessment run was executed in the same date, no log should be collected.

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

#### First execution specifying an only_logs_after value set in the past
Remove the database file and run the command specifying the date of the last assessment run. All events from that date onwards should be collected.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs are processed. Use the same `only_logs_after` parameter.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

#### Execute the command again with an only_logs_after value set earlier in the past

Run the module again, keeping the database file, using the date set for the first assessment run registered. The module should process no new log.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having a last key set and no only_logs_after value
Remove the database and execute the command `/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -p dev -r us-east-1 -s 2022-Jan-03`, which is the date of the first assessment run.
Then execute `sqlite3 ./aws_services.db 'delete from aws_services where scan_date != (SELECT min(scan_date) from aws_services);'`. This will simulate an environment which has only collected the logs corresponding to the first assessment run.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -p dev -r us-east-1
```
The module should collect all the logs starting from the `2022-Jan-03` without processing duplicates.

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having an only_logs_after later than the last key
Keeping the database file, execute `sqlite3 ./aws_services.db 'delete from aws_services where scan_date != (SELECT min(scan_date) from aws_services);'`.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```
Being `YYYY-MMM-DD` the date corresponding to the day following the execution date. No log should be collected.

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

#### Execute the module having an only_logs_after earlier than the last key
Keeping the database file, execute `sqlite3 ./aws_services.db 'delete from aws_services where scan_date != (SELECT min(scan_date) from aws_services);'`.
e
Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s 2021-Dec-01 -p dev -r us-east-1
```

It should only collect the logs starting from the first assessment run without processing it again.

<details>
    <summary>Command output</summary>


</details>

##### Conclusions


### Using reparse
Without removing the databases file, execute the following command:

```
/var/ossec/wodles/aws/aws-s3 -sr inspector -d 2 -s 2021-Jan-03 -p dev -r us-east-1 --reparse
```

<details>
    <summary>Command output</summary>


</details>

#### Conclusions

**Comment by CarlosRS9** [comment]
# Other bucket types and services

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Bucket/Service type | Test result | Issue ref. |
|---------------------|-------------|------------|
| CloudTrail          |             | -          |
| VPC                 |             | -          |
| Config              |             | -          |
| ALB                 |             | -          |
| CLB                 |             | -          |
| NLB                 |             | -          |
| KMS                 |             | -          |
| Macie               |             | -          |
| Trusted Advisor     |             | -          |
| GuardDuty           |             | -          |
| WAF                 |             | -          |
| Server Access       |             | -          |
| CloudWatch          |             | -          |

## Test details

### Test Cloudtrail

Remove the database file and run the command the using 2021-NOV-23 date as only_logs_after. 7 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-cloudtrail -t cloudtrail -s 2021-NOV-23 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

### Test VPCFlow

Remove the database file and run the command the using `2021-Dec-30` date as `only_logs_after`. 4 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -s 2021-Dec-30 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs have been processed.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-vpcflow -t vpcflow -s 2021-Dec-30 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Config

Remove the database file and run the command the using `2021-Dec-30` date as `only_logs_after`. 4 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -s 2021-Dec-30 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

Keep the database file and run the same command again to ensure no duplicate logs have been processed.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-config -t config -s 2021-Dec-30 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions

### Test ALB

Remove the database file and run the command the using 2020-Dec-21 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-alb -t alb -s 2020-Dec-21 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test CLB

Remove the database file and run the command the using 2021-NOV-11 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-clb -t clb -s 2021-NOV-11 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test NLB

Remove the database file and run the command the using 2021-NOV-11 date as only_logs_after. 5 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-nlb -t nlb -s 2020-NOV-21 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test KMS

Remove the database file and run the command the using 2021-NOV-19 date as only_logs_after. 6 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-kms -t custom -s 2021-NOV-19 -p dev -d2 | grep -v 'Skipping file with'
```
The output was piped to the `grep -v ...` command to avoid a large number of `DEBUG: ++ Skipping file with another prefix: test_prefix/kms_compress_encrypted/2018/....` lines. This issue will be addressed in #11148. The output of other commands will also be piped to grep to avoid those unnecessary lines.
<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Macie

Remove the database file and run the command the using 2021-NOV-11 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-macie -t custom -s 2021-NOV-11 -p dev -d2 | grep -v 'Skipping file with'
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Trusted Advisor

Remove the database file and run the command the using 2021-Dec-15 date as only_logs_after. 5 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-trusted-advisor -t custom -s 2021-Dec-15 -p dev -d2
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Guard Duty

Remove the database file and run the command the using 2021-Jun-22 date as only_logs_after. 3 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-guardduty -t custom -s 2021-Jun-22 -p dev -d2 | grep -v 'Skipping file with'
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test WAF

Remove the database file and run the command the using 2021-Dec-22 date as only_logs_after. 4 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-waf -t custom -s 2021-Sep-17 -p dev -d2 | grep -v 'Skipping file with'
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test Server Access

Remove the database file and run the command the using 2021-NOV-12 date as only_logs_after. 4 logs should be collected and sent to analysisd.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -b wazuh-aws-wodle-access -t server_access -s 2021-NOV-12 -p dev -d2 | grep -v 'Skipping file with'
```

<details>
    <summary>Command output</summary>

</details>

##### Conclusions


### Test CloudWatch Logs

Remove the database file and run the command specifying a date set 5 days earlier in the past. All the logs between the date set by the `only_logs_after` parameter and the present day must be collected.

Use the command 
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>


</details>

Keep the database file and run the same command again to ensure no duplicate logs are processed. Use the same `only_logs_after` parameter.

Use the command:
```
/var/ossec/wodles/aws/aws-s3 -sr cloudwatchlogs -g 4_3_test -d 2 -s YYYY-MMM-DD -p dev -r us-east-1
```

<details>
    <summary>Command output</summary>


</details>

##### Conclusions

**Comment by CarlosRS9** [comment]
# Agent testing

A test is considered valid if the following conditions are met:

- The expected result described in the test itself is met.
- The expected amount of events were sent to `analysisd`.
- The expected output was present in the `{wazuh_path}/logs/ossec.log` file.

The data available in our Wazuh-dev environment for each of these supported services will be used. Additional logs may be created if they are necessary or there is insufficient data for testing.

## Test results

| Test | Status | Issue ref. |
|---|---|---|
| Test the module in an agent |  | - |



## Test the module in an agent

Deploy an agent and install the required dependencies described in the official Wazuh documentation, then run the module adding the following configuration to the ossec.conf:

<details>
    <summary>Configuration</summary>

    <wodle name="aws-s3">
        <disabled>no</disabled>
        <remove_from_bucket>no</remove_from_bucket>
        <interval>10m</interval>
        <run_on_start>yes</run_on_start>
        <skip_on_error>no</skip_on_error>

        <bucket type="cloudtrail">
            <aws_profile>dev</aws_profile>
            <name>wazuh-aws-wodle-cloudtrail</name>
            <only_logs_after>2021-NOV-21</only_logs_after>
        </bucket>

        <bucket type="custom">
            <aws_profile>dev</aws_profile>
            <name>wazuh-aws-wodle-macie</name>
            <only_logs_after>2021-NOV-10</only_logs_after>
        </bucket>

        <service type="cloudwatchlogs">
            <aws_profile>dev</aws_profile>
            <aws_log_groups>4_3_test</aws_log_groups>
            <only_logs_after>2021-AUG-20</only_logs_after>
            <regions>us-east-1</regions>
        </service>
    </wodle>
</details>

<details>
    <summary>Output</summary>


</details>

In the Wazuh manager after enabling the `logall` option in the `ossec.conf` file, we can see that all the integrations were properly executed:
<details><summary>Cloudtrail</summary>


</details>

<details><summary>Macie</summary>

</details>

<details><summary>CloudWatch Logs</summary>


</details>

#### Conclusions
